contracts/abstract/AbstractNFTRewardDelegate.sol

```diff
@@ -0,0 +1,391 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.6;
+
+import '@openzeppelin/contracts/access/Ownable.sol';
+import '@openzeppelin/contracts/utils/Strings.sol';
+import {ERC721 as ERC721Rari} from '@rari-capital/solmate/src/tokens/ERC721.sol';
+
+import '@jbx-protocol/contracts-v2/contracts/interfaces/IJBDirectory.sol';
+import '@jbx-protocol/contracts-v2/contracts/interfaces/IJBFundingCycleDataSource.sol';
+import '@jbx-protocol/contracts-v2/contracts/interfaces/IJBPayDelegate.sol';
+import '@jbx-protocol/contracts-v2/contracts/interfaces/IJBRedemptionDelegate.sol';
+
+import '@jbx-protocol/contracts-v2/contracts/structs/JBDidPayData.sol';
+import '@jbx-protocol/contracts-v2/contracts/structs/JBDidRedeemData.sol';
+import '@jbx-protocol/contracts-v2/contracts/structs/JBRedeemParamsData.sol';
+import '@jbx-protocol/contracts-v2/contracts/structs/JBTokenAmount.sol';
+
+import '../interfaces/INFTRewardDataSourceDelegate.sol';
+import '../interfaces/IToken721UriResolver.sol';
+import '../interfaces/ITokenSupplyDetails.sol';
+
+/**
+  @title Juicebox data source delegate that offers project contributors NFTs.
+
+  @notice This contract allows project creators to reward contributors with NFTs. Intended use is to incentivize initial project support by minting a limited number of NFTs to the first N contributors.
+
+  @notice One use case is enabling the project to mint an NFT for anyone contributing any amount without a mint limit. Set minContribution.value to 0 and maxSupply to uint256.max to do this. To mint NFTs to the first 100 participants contributing 1000 DAI or more, set minContribution.value to 1000000000000000000000 (3 + 18 zeros), minContribution.token to 0x6B175474E89094C44Da98b954EedeAC495271d0F and maxSupply to 100.
+
+  @dev Keep in mind that this PayDelegate and RedeemDelegate implementation will simply pass through the weight and reclaimAmount it is called with.
+ */
+abstract contract AbstractNFTRewardDelegate is
+  ERC721Rari,
+  Ownable,
+  INFTRewardDataSourceDelegate,
+  IJBFundingCycleDataSource,
+  IJBPayDelegate,
+  IJBRedemptionDelegate
+{
+  using Strings for uint256;
+
+  //*********************************************************************//
+  // --------------------------- custom errors ------------------------- //
+  //*********************************************************************//
+  error INVALID_PAYMENT_EVENT();
+  error INCORRECT_OWNER();
+  error INVALID_ADDRESS();
+  error INVALID_TOKEN();
+  error SUPPLY_EXHAUSTED();
+  error NON_TRANSFERRABLE();
+  error INVALID_REQUEST(string);
+
+  //*********************************************************************//
+  // -------------------------- constructor ---------------------------- //
+  //*********************************************************************//
+
+  /**
+    @notice
+    Project id of the project this configuration is associated with.
+  */
+  uint256 private _projectId;
```

Comment:
please separate private and internal variable (+ are private really needed? Would you shadow-overwrite them in children contracts without?)